### PR TITLE
Replace C TLS shim with Rust OpenSSL wrapper

### PR DESCRIPTION
**Pull request by qbradley**
Rewrite kimojio-tls from the custom C OpenSSL shim to a Rust implementation over openssl-sys while preserving the existing BIO-pair driven TLS behavior. The new wrapper keeps SSL objects created through the openssl crate, removes the cc build step and C source, and makes the single-threaded SSL/BIO ownership model explicit in Rust.

Use Rc-owned TLS state so the wrapper is intentionally !Send/!Sync, expose BIO access through scoped synchronous operations instead of escaping raw BIO-backed slices, and keep debug-only reentrancy checks for accidental nested BIO access. This makes the formerly implicit safety invariants visible in the Rust type system without adding release-path reentrancy overhead.

Fix the SSL_read/SSL_write zero-return handling uncovered during the rewrite by classifying non-empty <= 0 returns through SSL_get_error, while preserving Success(0) for empty caller buffers. Add focused kimojio-tls regression tests for zero-return classification, length guards, clone/drop ownership smoke coverage, !Send/!Sync, and debug reentrancy checks.

Replace rcgen-based test certificate generation with OpenSSL X509/key builders and remove rcgen from the dependency graph. This drops a large transitive dependency set, including outdated windows-sys versions pulled in through rcgen, and bumps the workspace crates to 0.17.0 for the breaking kimojio-tls API cleanup.

### CODE REVIEW

File: kimojio-tls/src/lib.rs [hunk 1]
```diff
@@ -206,202 +145,480 @@ pub enum Response {
...
+            /// Gets the reference to SSL object.
+            pub fn get_ssl_raw(&self) -> *mut c_void {
+                self.inner.get_ssl_raw()
+            }
+
+            pub fn shutdown(&mut self) -> Response {
+                self.inner.shutdown()
+            }
+
+            pub fn read(&mut self, buffer: &mut [u8]) -> Response {
```
Comment (youyuanwu) [inline kimojio-tls/src/lib.rs:447]: The error return can match the SslStream in rust-openssl.
There are nice error wrappings but not exposed publicly.
I ported the Error type here for ktls usage: https://github.com/youyuanwu/rust-openssl-ktls/blob/987c38e3a943385efef9a0fd6b7a2b38c9ed34cf/openssl-ktls/src/error.rs#L44

Reusing ErrorStack struct made the error conversion simpler. 
Maybe kimojio should do the same.

File: kimojio-tls/src/lib.rs [hunk 2]
```diff
@@ -1,35 +1,30 @@
...
-struct RawTlsServerContext {
-    _data: [u8; 0],
-    _marker: core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
-}
-
-#[repr(C)]
-struct RawError {
-    error_type: i32,
-    error_code: i32,
+unsafe extern "C" {
```
Comment (youyuanwu) [inline kimojio-tls/src/lib.rs:18]: We should document we only support openssl 3+? openssl crate has complicated logic to defined extern fn from different openssl versions.

File: kimojio/src/tlscontext.rs [hunk 3]
```diff
@@ -51,11 +48,39 @@ impl TlsContext {
...
         client.client_side_handshake(deadline).await?;
         Ok(client)
     }
+
+    fn create_ssl(&self, bufsize: usize, is_server: bool) -> Result<TlsServer, Errno> {
+        let ssl = openssl::ssl::Ssl::new(&self.ssl_ctx).map_err(as_openssl_error)?;
+        let raw_ssl = ssl.as_ptr() as *mut std::ffi::c_void;
+        std::mem::forget(ssl);
+        // SAFETY: `raw_ssl` came from a freshly-created `openssl::ssl::Ssl`.
+        // We forgot the Rust owner, transferring ownership to kimojio_tls.
```
Comment (youyuanwu) [inline kimojio/src/tlscontext.rs]: Should we always use openssl::ssl::Ssl everywhere to make the ownership clear? kimojio_tls can take the moved value.

File: kimojio/src/tlsstream.rs [hunk 4]
```diff
@@ -167,18 +236,22 @@ impl SplittableStream for TlsStream {
...
+    ssl: &mut impl TlsIo,
     socket: &impl SocketPair,
     deadline: Option<Instant>,
 ) -> Result<(), Errno> {
     let socket = socket.read_socket().await?;
     if let Some(socket) = socket.borrow() {
-        if let Some(buffer) = ssl.get_push_buffer() {
-            let amount = operations::read_with_deadline(socket, buffer, deadline).await?;
+        if let Some(capacity) = ssl.push_buffer_capacity() {
+            let mut buffer = vec![0; capacity];
```
Comment (youyuanwu) [inline kimojio/src/tlsstream.rs]: Should this vec be saved in TlsIo to avoid allocation each time?

File: kimojio/src/tlsstream.rs [hunk 5]
```diff
@@ -197,8 +270,8 @@ async fn try_write(
         // buffer by the length of the full buffer copied into the BIO rather than
         // the amount of data written to the socket here. Need to fully exhaust
         // the BIO here to keep the bookkeeping in sync
-        while let Some(buffer) = ssl.get_pull_buffer() {
-            let amount = operations::write_with_deadline(socket, buffer, deadline).await?;
+        while let Some(buffer) = ssl.copy_pull_buffer() {
```
Comment (youyuanwu) [inline kimojio/src/tlsstream.rs]: Does this require additional copy compared to the original code?

File: kimojio/src/tlscontext.rs (hunk 3 above)
Comment (qbradley) [reply kimojio/src/tlscontext.rs]: I think this would be a lot cleaner.  I will give it a try. It will move us farther from the C implementation and exposes us to possibility of overheads from openssl crate but it would be nice if it could work out.

File: kimojio/src/tlsstream.rs (hunk 4 above)
Comment (qbradley) [reply kimojio/src/tlsstream.rs]: I was trying to make BIO use safer but I don't want the copy. I think I'll need something more like RefCell than a callback I was trying to use.

File: kimojio-tls/src/lib.rs (hunk 1 above)
Comment (qbradley) [reply kimojio-tls/src/lib.rs:447]: fixed

File: kimojio-tls/src/lib.rs (hunk 2 above)
Comment (qbradley) [reply kimojio-tls/src/lib.rs:18]: These actually do work with OpenSSL 1.  They won't work with BoringSSL or some other options for openssl-sys crate, however in those cases you will get a link error, so I think it is ok.

File: kimojio/src/tlscontext.rs (hunk 3 above)
Comment (qbradley) [reply kimojio/src/tlscontext.rs]: fixed

File: kimojio/src/tlsstream.rs (hunk 4 above)
Comment (qbradley) [reply kimojio/src/tlsstream.rs]: fixed

File: kimojio/src/tlsstream.rs (hunk 5 above)
Comment (qbradley) [reply kimojio/src/tlsstream.rs]: fixed

File: kimojio-tls/src/lib.rs [hunk 6]
```diff
@@ -206,202 +139,542 @@ pub enum Response {
...
+}
+
+impl Drop for ApplicationBioCallbackGuard<'_> {
     fn drop(&mut self) {
-        unsafe {
-            tls_handle_ctx_close(self.ctx);
+        match self.direction {
+            ApplicationBioDirection::Read => self.bio.in_read_callback.set(false),
+            ApplicationBioDirection::Write => {
+                #[cfg(debug_assertions)]
```
Comment (youyuanwu) [inline kimojio-tls/src/lib.rs]: Why write is under debug-mode and read is not?

File: kimojio-tls/src/lib.rs [hunk 7]
```diff
@@ -206,202 +139,542 @@ pub enum Response {
...
+impl Drop for TlsServerInner {
+    fn drop(&mut self) {}
+}
 
-    pub fn read(&mut self, buffer: &mut [u8]) -> Response {
-        get_response(unsafe {
-            tls_handle_read(self.server, buffer.as_mut_ptr(), buffer.len() as isize)
+/// Gets the minimum TLS protocol version from an OpenSSL context.
+pub fn get_min_proto_version(ctx: &SslContextRef) -> i32 {
+    unsafe { ffi::SSL_CTX_get_min_proto_version(ctx.as_ptr()) }
```
Comment (youyuanwu) [inline kimojio-tls/src/lib.rs:370]: There might be a safe function already to get version.

**Review by youyuanwu — APPROVED** [review]

File: kimojio-tls/src/lib.rs (hunk 6 above)
Comment (qbradley) [reply kimojio-tls/src/lib.rs]: fixed

File: kimojio-tls/src/lib.rs (hunk 7 above)
Comment (qbradley) [reply kimojio-tls/src/lib.rs:370]: There is, but it is not conveniently exposed.

**Review by youyuanwu — APPROVED** [review]
